### ISSUE TITLE
[MPS] `torch.nn.functional.conv_transpose3d` return unexpected results when weight containing `inf`

### ISSUE THREAD

**Issue by can-gaa-hou** [pytorch/pytorch#164125]
### 🐛 Describe the bug

Similar to #156062 (cannot reproduce now, may be fixed), `torch.nn.functional.conv_transpose3d` returns all `nan` when the weight contains `inf` on MPS.

```python
import torch
input = torch.zeros((1, 1, 5, 5, 5), dtype=torch.float32)
weight = torch.zeros((1, 1, 3, 3, 3), dtype=torch.float32)
weight[0, 0, 0, 0, 0] = float('inf')

input_mps = input.to("mps")
weight_mps = weight.to("mps")
out_mps = torch.nn.functional.conv_transpose3d(input=input_mps, weight=weight_mps)
out_cpu = torch.nn.functional.conv_transpose3d(input=input, weight=weight)
print(out_cpu) # contains zero
print("--------------------------------")
print(out_mps.cpu()) # all nan
``` 

### Versions

nightly built

**Comment by jhavukainen** [comment]
Looked into this a little bit as well as the proposed fix above. Some notes on that:

1. Both the CPU and MPS behaving correctly even though the outputs differ as there is no formal definition in what is the expected propagation of `nan` when one weight element contains an `inf`. The reason why CPU only has some of the elements set to `nan` and others to zero while MPSGraph outputs all `nan` has to do with how the computation is performed: CPU side seems to do a bounds check for the receptive-field when scattering the results limiting the `nan`s to only the corner where the `inf` weight was. Based on the symptoms the MPS implementation differs and possibly uses some zero-padding trick to perform the computation more efficiently but since `0 * inf = nan` the nonfinite values propagate differently.
2. While the CPU implementation is often used to provide a reference output for checking correctness there's no formal requirement for each backend to consider it as the ground truth, especially as the implementations there are not infallible either. But as said in this case both are correct they just implement the convolution algorithm differently. Notably doing similar testing for the forward pass also shows that the CPU propagates the inf/nan in a "suprising" way that's not consistent between the different CPU implementations (winograd vs. slow2d) since there's no real reference how it needs to be done.

So I'd say to conclusion here is that as there's nothing wrong happening we should not probably fix it. The issue is already in the model if one of the weights has turned into `inf` so adding complications to force it to propagate similarly to the CPU implementations is probably not worthwhile.

That being said if there is a pressing edict to make this behavior match the CPU, the least harmful way to do it would be to check if any of the weights is inf on the host side and split the graph caching into two cases: one for inf weight graphs and one for regular cases (the existing path). This is a bit surprising but after testing around the idea for a while and trying to do the work within the MPSGraph as suggested in https://github.com/pytorch/pytorch/pull/169238 it looks like there's only bad options as doing a "bounds check" of sort for the receptive field will add overhead to all calls and using the control flow operations within graph to try avoiding that will underneath also force the sync with the CPU to make the call and again adds overhead to all invocations.

**Comment by can-gaa-hou** [comment]
Thanks for the detailed investigation and explanation. @jhavukainen 

The reasoning about the different nan propagation behavior due to the implementation differences between CPU and MPSGraph makes sense to me. I also agree that this falls more into implementation-defined behavior rather than a correctness issue, and that adding overhead to the normal execution path just to match CPU propagation semantics probably would not be worthwhile.

Appreciate you digging into this and explaining the tradeoffs in detail. I would like to close this issue as done.
